### PR TITLE
COVER-5: Add new command to re-build local filesystem

### PR DESCRIPTION
**Pull request by cableman**
#### Link to ticket

https://jira.itkdev.dk/browse/COVER-5

#### Description

- Commands clean-up
- New rebuild filesystem by downloading covers from cover store

#### Screenshot of the result

N/A

#### Checklist

- [x] My code passes our test (all our tests).
- [x] My code passes our static analysis suite.
- [x] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change 
should be exempt from the list.

#### Additional comments or questions

N/A

### CODE REVIEW

File: src/Command/CleanUpCommand.php [hunk 1]
```diff
@@ -51,15 +52,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+                    if (null !== $item) {
+                        $cover->setRemoteUrl($item->getUrl());
+                    }
+                    $cover->setUploaded(true);
+                    $this->entityManager->flush();
                 }
-                $cover->setUploaded(true);
-                $this->entityManager->flush();
+            } catch (UninitializedPropertyException $e) {
+                // Don't do anything as we have covers without material.
```
Comment (turegjorup) [inline src/Command/CleanUpCommand.php:67]: Not obvious where the exception is throw. Also not fan of handling this with `try/catch`. Exception handling should be for actual exceptions. If a cover legitimately doesn't need to have a `Material` then this is not an exception. If we don't want to return `?Material` and do a `null` check then `Cover` should have a `hasMaterial(): bool`.

File: src/Command/CleanUpCommand.php [hunk 2]
```diff
@@ -51,15 +52,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
         $query = $this->coverRepository->getIsNotUploadedQuery($limit);
         /** @var Cover $cover */
         foreach ($query->toIterable() as $cover) {
-            if ($this->coverStoreService->exists($cover->getMaterial()->getIsIdentifier())) {
-                $this->coverStoreService->removeLocalFile($cover);
+            try {
+                if ($this->coverStoreService->exists($cover->getMaterial()->getIsIdentifier())) {
+                    $this->coverStoreService->removeLocalFile($cover);
```
Comment (turegjorup) [inline src/Command/CleanUpCommand.php:57]: I know this is only a refactor. But I really think it was the wrong decision to not keep a local copy of the file. A CDN is not file storage.  As the need for this PR clearly demonstrates.

Would prefer we remove this command.

File: src/Command/CleanUpDatabaseCommand.php [hunk 3]
```diff
@@ -60,7 +61,9 @@ protected function execute(InputInterface $input, OutputInterface $output): int
         foreach ($query->toIterable() as $cover) {
             $output->write('.');
 
-            if (is_null($cover->getMaterial())) {
+            try {
+                $cover->getMaterial();
+            } catch (UninitializedPropertyException $e) {
```
Comment (turegjorup) [inline src/Command/CleanUpDatabaseCommand.php:66]: See above. Exceptions shouldn't be (mis)used for handling normal flow.

File: src/Command/ExistsCommand.php [hunk 4]
```diff
@@ -47,26 +46,31 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
-                if (!is_null($material->getCover())) {
+                try {
+                    $material->getCover();
                     if ($this->coverStoreService->exists($material->getIsIdentifier())) {
                         $output->write((string) $material);
                         $inCoverServiceIndex = true;
                     }
                     $inCoverTable = true;
+                } catch (UninitializedPropertyException $e) {
+                    $noCover = true;
```
Comment (turegjorup) [inline src/Command/ExistsCommand.php:65]: See above

File: src/Command/GetRemoteUrlsCommand.php [hunk 5]
```diff
@@ -54,15 +55,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
+                    if (null !== $item) {
+                        $cover->setRemoteUrl($item->getUrl());
+                    }
+                    $cover->setUploaded(true);
+                    $this->entityManager->flush();
                 }
-                $cover->setUploaded(true);
-                $this->entityManager->flush();
+            } catch (UninitializedPropertyException $e) {
+                // Do nothing, as cover was not connect to a material.
```
Comment (turegjorup) [inline src/Command/GetRemoteUrlsCommand.php:70]: See above

File: src/Command/GetRemoteUrlsCommand.php [hunk 6]
```diff
@@ -54,15 +55,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 
         /** @var Cover $cover */
         foreach ($query->toIterable() as $cover) {
-            if ($this->coverStoreService->exists($cover->getMaterial()->getIsIdentifier())) {
-                $this->coverStoreService->removeLocalFile($cover);
-                $item = $this->coverStoreService->search($cover->getMaterial()->getIsIdentifier());
+            try {
+                if ($this->coverStoreService->exists($cover->getMaterial()->getIsIdentifier())) {
+                    $this->coverStoreService->removeLocalFile($cover);
```
Comment (turegjorup) [inline src/Command/GetRemoteUrlsCommand.php:60]: See above about local files

File: src/Command/RebuildFilesCommand.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+        /** @var Cover $cover */
+        foreach ($query->toIterable() as $cover) {
+            if (!$this->coverStoreService->existsLocalFile($cover)) {
+                $remoteUrl = $cover->getRemoteUrl();
+                if (!is_null($remoteUrl)) {
+                    $response = $this->httpClient->request('GET', $remoteUrl);
+                    if (200 !== $response->getStatusCode()) {
+                        // Download failed.
+                        $output->write('E');
```
Comment (turegjorup) [inline src/Command/RebuildFilesCommand.php:65]: Not sure this is sufficient output for debugging?

**Review by turegjorup — CHANGES_REQUESTED** [review]

File: src/Command/CleanUpCommand.php (hunk 1 above)
Comment (cableman) [reply src/Command/CleanUpCommand.php:67]: It's an side effect of the way you first upload a cover and then in another request link it to material.

**Pull request merged**